### PR TITLE
Studio editor validation error UI improvements

### PR DESCRIPTION
**Pull request by nsprenkle**
**TL;DR -** Several clarity and quality of life improvements for studio editor frontend validation error workflows.

JIRA: [AU-37](https://openedx.atlassian.net/browse/AU-37) & [AU-63](https://openedx.atlassian.net/browse/AU-63)

**What changed?**

- Change validation banner text to list tabs which fail validation
- Add warning icons next to tabs which fail validation
- Larger, centered validation banner close icon
- Backend refactors to support the above including:
    - Adding metadata to link views and tabs.
    - Validation of assessment step dates moved to schedule tab, where the selectors live. 

**Developer Checklist**

- [x] Reviewed the [release process](https://github.com/edx/edx-ora2/blob/master/.github/release_process.md)
- [x] Translations and JS/SASS compiled
- [x] Bumped version number in [setup.py](https://github.com/edx/edx-ora2/blob/a62e81a9b0d89223476967ec3c27f3557a850735/setup.py#L39) and [package.json](https://github.com/edx/edx-ora2/blob/a62e81a9b0d89223476967ec3c27f3557a850735/package.json#L3)

Example validation errors for Rubric and Assessment Steps tabs:
Example validation errors for Rubric and Assessment Steps tabs

**Testing Instructions**

1. In Studio, visit or create an ORA.
2. Click "EDIT"
3. Introduce validation errors on any combination of the below tabs, using your own or example invalid cases provided below.
    a. Prompt - always valid
    b. Rubric - remove all criteria
    c. Schedule - enter an invalid date for assignment or assessment step
    d. Assessment Steps - enter NaN for the "Must Grade" in Peer Learning step
    e. Settings - set file uploads to required/optional, custom allow-list, but remove all file types from the custom allow-list
4. Validate that invalid tabs are:
    1. Marked with an invalid, ⚠️ icon
    2. Show up in the "Save Unsuccessful" error banner list
5. Validate that valid tabs are:
    1. Not marked as invalid
    2. Do not show up in the "Save Unsuccessful" error banner list
6. General smoke testing of save/edit functionality

**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:

- [ ] I've done a visual code review
- [ ] I've tested the new functionality

FYI: @edx/masters-devs-gta

### CODE REVIEW

File: openassessment/xblock/static/js/src/studio/oa_edit_settings.js [hunk 1]
```diff
@@ -21,7 +21,6 @@ Returns:
 * */
 export class EditSettingsView {
   constructor(element, assessmentViews, data) {
-    const self = this;
```
Comment (leangseu-edx) [inline openassessment/xblock/static/js/src/studio/oa_edit_settings.js:24]: 👍

File: openassessment/xblock/static/js/src/studio/oa_edit.js [hunk 2]
```diff
@@ -161,12 +161,16 @@ export class StudioView {
...
     this.clearValidationErrors();
-    if (!this.validate()) {
+    const tabsFailingValidation = this.validate();
+    this.markTabsWithValidationErrors(tabsFailingValidation);
+
+    if (tabsFailingValidation.length > 0) {
+      const tabNames = tabsFailingValidation.map(tab => tab.tabDisplayName);
       this.alert.setMessage(
-        gettext('Couldn\'t Save This Assignment'),
-        gettext('Please correct the outlined fields.'),
```
Comment (leangseu-edx) [inline openassessment/xblock/static/js/src/studio/oa_edit.js:169]: Should we preserve this message?

**Review by leangseu-edx — CHANGES_REQUESTED** [review]
Create an `open assessment` then r**emove all criterion** and **delete one of the date** in schedule tab. The icon show up in the wrong tab.

image

File: openassessment/xblock/static/js/src/studio/oa_edit_assessments_steps.js [hunk 3]
```diff
@@ -17,6 +17,12 @@ export class EditAssessmentsStepsView {
     this.initializeSortableAssessments();
   }
 
+  // Expose tab name for error messaging
+  tabDisplayName = gettext('Assessment Steps');
```
[the remark this answers is not in the document]
Comment (nsprenkle) [reply openassessment/xblock/static/js/src/studio/oa_edit_assessments_steps.js:21]: Possibly, but I figured this was way cleaner and less fragile than trying to figure out how to rip these dynamically from the DOM.

File: openassessment/xblock/static/js/src/studio/oa_edit.js (hunk 2 above)
Comment (nsprenkle) [reply openassessment/xblock/static/js/src/studio/oa_edit.js:169]: This was an intentional language change from UX so I'm going to say we leave it as coded.

**Comment by nsprenkle** [comment]
> Create an `open assessment` then r**emove all criterion** and **delete one of the date** in schedule tab. The icon show up in the wrong tab.
> 
> image

@leangseu-edx, you're going to love this. If you clear the "Course deadlines" date you'll see it correctly attribute to the "Schedule" tab. What I didn't realize is that the assessment step dates (e.g. "Self Assessment Deadlines" as you edited) are validated as part of the "Assessment Steps" even though they exist on the schedule tab.

I have no idea how to cleanly delineate these...

File: openassessment/xblock/static/js/src/studio/oa_edit.js [hunk 4]
```diff
@@ -268,19 +272,45 @@ export class StudioView {
...
+      this.markTabAsInvalid(tab.tabTarget);
+    });
+  }
+
+  /**
+   * Given an aria controls target, add or remove validation warning styling.
+   *  value - true enables invalid styling, false removes invalid styling
+   * */
+  markTabAsInvalid(controlTarget, value = true) {
+    const tab = $(`[aria-controls='${controlTarget}']`, this.element);
```
[the remark this answers is not in the document]
Comment (nsprenkle) [reply openassessment/xblock/static/js/src/studio/oa_edit.js:312]: I like that the `href` approach is more semantically accurate but it is also more fragile since it becomes inherently more tied to the DOM structure. I'm thinking IDs are the way to go but, candidly, all of this is a cluster...

**Comment by jansenk** [comment]
> @leangseu-edx, you're going to love this. If you clear the "Course deadlines" date you'll see it correctly attribute to the "Schedule" tab. What I didn't realize is that the assessment step dates (e.g. "Self Assessment Deadlines" as you edited) are validated as part of the "Assessment Steps" even though they exist on the schedule tab.

Step deadlines originally were on the assessment steps page. They were then later added to the schedules tab. My guess is that when that move happened whoever did it didn’t make the corresponding change in validation. My vote is to open up a new bug ticket

File: openassessment/xblock/static/js/src/studio/oa_edit_assessments_steps.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (nsprenkle) [reply openassessment/xblock/static/js/src/studio/oa_edit_assessments_steps.js]: Actually, hopefully killed 2 ~birds~ complaints with one ~stone~ commit (4280fe8): adding IDs to tabs means we can reliably link to tabs and get their text directly from translated template.

File: openassessment/xblock/static/dist/95ec738c0b7faac5b5c9126794446bbd.svg [hunk 5]
```diff
@@ -0,0 +1,4 @@
+<svg width="36" height="36" viewBox="0 0 36 36" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (leangseu-edx) [inline openassessment/xblock/static/dist/95ec738c0b7faac5b5c9126794446bbd.svg:1]: I do not think this file should be included.

**Review by leangseu-edx — APPROVED** [review]
Beside the `svg` file, this is g2g.

**Comment by nsprenkle** [comment]
> Beside the `svg` file, this is g2g.

Actually, I think it should be here (it's the default avatar image). It keeps showing up in builds and is actually included in the build manifest. _Something_ clearly is adding it so I think we just forgot to add it as part of the build.
